tests/AutoReview/ProjectCode/ProjectCodeProvider.php

```diff
@@ -0,0 +1,292 @@
+use Infection\TestFramework\Coverage\CoverageLineData;
+use Infection\TestFramework\Coverage\CoverageMethodData;
+use Infection\TestFramework\Coverage\TestFileTimeData;
+use Infection\TestFramework\PhpSpec\Config\Builder\InitialConfigBuilder as PhpSpecInitalConfigBuilder;
+use Infection\TestFramework\PhpSpec\Config\Builder\MutationConfigBuilder as PhpSpecMutationConfigBuilder;
+use Infection\TestFramework\PhpSpec\Config\NoCodeCoverageException;
+use Infection\TestFramework\PhpUnit\Config\Builder\InitialConfigBuilder as PhpUnitInitalConfigBuilder;
+use Infection\TestFramework\PhpUnit\Config\Builder\MutationConfigBuilder as PhpUnitMutationConfigBuilder;
+use Infection\TestFramework\PhpUnit\Coverage\CoverageXmlParser;
+use Infection\TestFramework\TestFrameworkTypes;
+use function Infection\Tests\generator_to_phpunit_data_provider;
+use Infection\Utils\VersionParser;
+use Infection\Visitor\MutationsCollectorVisitor;
+use Infection\Visitor\ParentConnectorVisitor;
+use function iterator_to_array;
+use ReflectionClass;
+use const SORT_STRING;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+
+final class ProjectCodeProvider
+{
+    /**
+     * List of classes which should not have a matching test.
+     */
+    public const NON_TESTED_CONCRETE_CLASSES = [
+        ConfigureCommand::class,
+        InfectionCommand::class,
+        Application::class,
+        ProgressFormatter::class,
+        PhpProcess::class,
+        ComposerExecutableFinder::class,
+        BadgeApiClient::class,
+        ResultsLoggerTypes::class,
+        MutantCreatingConsoleLoggerSubscriber::class,
+        MutationGeneratingConsoleLoggerSubscriber::class,
+        MutationTestingRunner::class,
+        NoCodeCoverageException::class,
+        TestFrameworkTypes::class,
+        MutationsCollectorVisitor::class,
+        ParentConnectorVisitor::class,
+    ];
+
+    /**
+     * This array contains all classes that are not extension points, but not final due to legacy
+     * reasons. This list should never be added to, only removed from.
+     */
+    public const NON_FINAL_EXTENSION_CLASSES = [
+        ConsoleHelper::class,
+        SourceDirGuesser::class,
+        InfectionConfig::class,
+        DiffColorizer::class,
+        Differ::class,
+        TestFrameworkFinder::class,
+        BadgeApiClient::class,
+        MetricsCalculator::class,
+        ProcessBuilder::class,
+        CodeCoverageData::class,
+        PhpSpecInitalConfigBuilder::class,
+        PhpUnitInitalConfigBuilder::class,
+        PhpSpecMutationConfigBuilder::class,
+        PhpUnitMutationConfigBuilder::class,
+        CoverageXmlParser::class,
+        VersionParser::class,
+    ];
+
+    /**
+     * This array contains all classes that can be extended by our users.
+     */
+    public const EXTENSION_POINTS = [
+        Mutator::class,
+        OutputFormatter::class,
+    ];
+
+    private static $sourceClasses;
```

Comment:
It would be nice for these 2 to have a doc comment, for type hinting etc.

Reply:
I updated the type, as to the doc comment I think it's unwise to do so since there usage are fairly internal and documenting how a property is used is... always gonna out of sync with the underlying code resulting in a false hence potentially confusing doc. The code usage speaks for itself IMO, and it's not a case that requires to go to great length to document